pkg/volume/flexvolume/probe_test.go

```diff
@@ -0,0 +1,293 @@
+
+	// Assert
+	assert.Equal(t, 3, len(watcher.watches)) // 2 from initial setup, 1 from new watch.
+	assert.Equal(t, pluginDir, watcher.watches[len(watcher.watches)-1])
+}
+
+// Issue multiple events and probe multiple times. Should give true, false, false...
+func TestProberMultipleEvents(t *testing.T) {
+	// Arrange
+	_, fs, watcher, prober := initTestEnvironment(t)
+	for i := 0; i < 5; i++ {
+		newDriver := fmt.Sprintf("multi-event-driver%d", 1)
+		installDriver(newDriver, fs)
+		driverPath := path.Join(pluginDir, newDriver)
+		watcher.TriggerEvent(fsnotify.Create, driverPath)
+		watcher.TriggerEvent(fsnotify.Create, path.Join(driverPath, newDriver))
+	}
+
+	// Act
+	updated, _, err := prober.Probe()
+
+	// Assert
+	assert.True(t, updated)
+	assert.NoError(t, err)
+	for i := 0; i < 4; i++ {
+		updated, _, err = prober.Probe()
+		assert.False(t, updated)
+		assert.NoError(t, err)
+	}
+}
+
+// When many events are triggered quickly in succession, events should stop triggering a probe update
+// after a certain limit.
+func TestProberRateLimit(t *testing.T) {
+	// Arrange
+	driverPath, _, watcher, prober := initTestEnvironment(t)
+	for i := 0; i < watchEventLimit; i++ {
+		watcher.TriggerEvent(fsnotify.Write, path.Join(driverPath, driverName))
+	}
+
+	// Act
+	updated, plugins, err := prober.Probe()
+
+	// Assert
+	assert.True(t, updated)
+	assert.Equal(t, 1, len(plugins))
+	assert.NoError(t, err)
+
+	// Arrange
+	watcher.TriggerEvent(fsnotify.Write, path.Join(driverPath, driverName))
+
+	// Act
+	updated, _, err = prober.Probe()
+
+	// Assert
+	assert.False(t, updated)
+	assert.NoError(t, err)
+}
+
+func TestProberError(t *testing.T) {
+	fs := utilfs.NewFakeFs()
+	watcher := NewMockWatcher()
+	prober := &flexVolumeProber{
+		pluginDir: pluginDir,
+		watcher:   watcher,
+		fs:        fs,
+		factory:   mockPluginFactory{error: true},
+	}
+	installDriver(driverName, fs)
+	prober.Init()
+
+	_, _, err := prober.Probe()
+	assert.Error(t, err)
+}
+
+// Installs a mock driver (an empty file) in the mock fs.
+func installDriver(driverName string, fs utilfs.Filesystem) {
+	driverPath := path.Join(pluginDir, driverName)
+	fs.MkdirAll(driverPath, 0666)
+	fs.Create(path.Join(driverPath, driverName))
+}
+
+// Initializes mocks, installs a single driver in the mock fs, then initializes prober.
+func initTestEnvironment(t *testing.T) (
+	driverPath string,
+	fs utilfs.Filesystem,
+	watcher *mockWatcher,
+	prober volume.DynamicPluginProber) {
+	fs = utilfs.NewFakeFs()
+	watcher = NewMockWatcher()
+	prober = &flexVolumeProber{
+		pluginDir: pluginDir,
+		watcher:   watcher,
+		fs:        fs,
+		factory:   mockPluginFactory{error: false},
+	}
+	driverPath = path.Join(pluginDir, driverName)
+	installDriver(driverName, fs)
+	prober.Init()
+
+	assert.NotNilf(t, watcher.eventHandler,
+		"Expect watch event handler to be registered after prober init, but is not.")
+	return
+}
+
+// Mock filesystem watcher
+type mockWatcher struct {
```

Comment:
Move this to separate file?

Reply:
+1. I'd also rename it to fakeWatcher, which is more conventional in our codebase

Reply:
ACK

Reply:
Keeping the file inside Flexvolume directory since fakes may look different depending on test.